### ISSUE TITLE
Using with system wide cxsparse

### ISSUE THREAD

**Issue by epsilon-0** [igraph/igraph#1481]
Make sure that these boxes are checked before submitting your issue -- thank you!

- [x] This issue is for the C core of igraph.
- [x] This issue is a bug report or a feature request, not a support question.

---

I've added a patch here which makes it possible to unbundle the shipped cxsparse and use the system-wide one instead.
This should make code maintenance simpler and also allows us to user newer versions.

This is currently used in Gentoo and it is working nicely, all tests are passing

Would like to get this in so that the package can become simpler and easier to maintain.

```
diff --git a/configure.ac b/configure.ac
index b422caf..dae8647 100644
--- a/configure.ac
+++ b/configure.ac
@@ -333,6 +333,8 @@ else
   internal_glpk=no
 fi
 
+PKG_CHECK_MODULES([CS],[cxsparse])
+
 # Link time optimization feature in newer gcc/g++
 # based on http://svn.r-project.org/R/trunk/configure.ac
 AC_ARG_ENABLE([lto],
diff --git a/examples/simple/igraph_sparsemat2.c b/examples/simple/igraph_sparsemat2.c
index 24877c2..0f4b45c 100644
--- a/examples/simple/igraph_sparsemat2.c
+++ b/examples/simple/igraph_sparsemat2.c
@@ -21,7 +21,7 @@
 
 */
 
-#include <cs/cs.h>
+#include <cs.h>
 #include <igraph.h>
 #include "igraph_blas_internal.h"
 #include "igraph_arpack_internal.h"
diff --git a/examples/simple/igraph_sparsemat3.c b/examples/simple/igraph_sparsemat3.c
index 09329f0..31d4f1f 100644
--- a/examples/simple/igraph_sparsemat3.c
+++ b/examples/simple/igraph_sparsemat3.c
@@ -21,7 +21,7 @@
 
 */
 
-#include <cs/cs.h>
+#include <cs.h>
 #include <igraph.h>
 
 int permute(const igraph_matrix_t *M,
diff --git a/examples/simple/igraph_sparsemat4.c b/examples/simple/igraph_sparsemat4.c
index 3a33979..c0ea489 100644
--- a/examples/simple/igraph_sparsemat4.c
+++ b/examples/simple/igraph_sparsemat4.c
@@ -21,7 +21,7 @@
 
 */
 
-#include <cs/cs.h>
+#include <cs.h>
 #include <igraph.h>
 
 igraph_bool_t check_solution(const igraph_sparsemat_t *A,
diff --git a/src/Makefile.am b/src/Makefile.am
index f9d02d2..8ab3385 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -9,9 +9,6 @@ AM_YFLAGS = -d
 
 lib_LTLIBRARIES         = libigraph.la
 
-include lapack/blas.inc
-include lapack/lapack.inc
-include lapack/arpack.inc
 include plfit/plfit.inc
 
 F2C =	f2c/abort_.c	  f2c/dolio.c	    f2c/r_sin.c\
@@ -99,8 +96,6 @@ if INTERNAL_ARPACK
   ARPACK_LIB = libarpack.la
 endif
 
-include ../optional/glpk/glpk.inc
-
 if INTERNAL_GLPK
   libglpk_la_SOURCES = 	$(GLPK)
   libglpk_la_CFLAGS = -I$(top_srcdir)/optional/glpk
@@ -365,7 +360,7 @@ SOURCES = 		     basic_query.c games.c cocitation.c iterators.c \
 			     drl_layout_3d.cpp drl_graph_3d.cpp \
 			     DensityGrid_3d.cpp \
 			     foreign-dl-parser.y foreign-dl-lexer.l \
-			     $(CS) sparsemat.c mixing.c bigint.c bignum.c \
+			     sparsemat.c mixing.c bigint.c bignum.c \
 			     version.c optimal_modularity.c \
 			     igraph_fixed_vectorlist.c separators.c \
 			     igraph_marked_queue.c igraph_estack.c st-cuts.c \
@@ -408,7 +403,7 @@ libigraph_la_CXXFLAGS	   = -I$(top_srcdir)/include -I$(top_builddir)/include $(W
 libigraph_la_LDFLAGS       = -no-undefined
 libigraph_la_LIBADD        = -lm $(XML2_LIBS) $(F2C_LIB) $(BLAS_LIB) \
 				 $(LAPACK_LIB) $(ARPACK_LIB) $(GLPK_LIB) $(PRPACK_LIB) \
-				 $(PLFIT_LIB)
+				 $(PLFIT_LIB) $(CS_LIBS)
 
 if INTERNAL_GLPK
   libigraph_la_CFLAGS   += -I$(top_srcdir)/optional/glpk
diff --git a/src/Makefile.in b/src/Makefile.in
index 56703d4..cc30d48 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -2411,7 +2411,7 @@ SOURCES = basic_query.c games.c cocitation.c iterators.c \
 	gengraph_powerlaw.cpp gengraph_random.cpp decomposition.c \
 	bipartite.c drl_layout_3d.cpp drl_graph_3d.cpp \
 	DensityGrid_3d.cpp foreign-dl-parser.y foreign-dl-lexer.l \
-	$(CS) sparsemat.c mixing.c bigint.c bignum.c version.c \
+	sparsemat.c mixing.c bigint.c bignum.c version.c \
 	optimal_modularity.c igraph_fixed_vectorlist.c separators.c \
 	igraph_marked_queue.c igraph_estack.c st-cuts.c \
 	cohesive_blocks.c statusbar.c lapack.c complex.c eigen.c \
diff --git a/src/sparsemat.c b/src/sparsemat.c
index 0ed4678..4fb724d 100644
--- a/src/sparsemat.c
+++ b/src/sparsemat.c
@@ -23,7 +23,7 @@
 
 #include "config.h"
 
-#include "cs/cs.h"
+#include <cs.h>
 
 #include "igraph_sparsemat.h"
 #include "igraph_error.h"
diff --git a/tests/Makefile.am b/tests/Makefile.am
index d350f11..31a1f5a 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -13,7 +13,7 @@ $(srcdir)/package.m4: $(top_srcdir)/configure.ac
 	  echo 'm4_define([AT_PACKAGE_BUGREPORT], [@PACKAGE_BUGREPORT@])'; \
 	} >$(srcdir)/package.m4
 
-EXTRA_DIST += package.m4
+#EXTRA_DIST += package.m4
 
 TESTSUITE_AT = \
 	testsuite.at \
diff --git a/tests/mt.at b/tests/mt.at
index 050974a..b154ce1 100644
--- a/tests/mt.at
+++ b/tests/mt.at
@@ -23,11 +23,11 @@ AT_BANNER([[Thread-safety tests]])
 
 AT_SETUP([Simple error handling test :])
 AT_KEYWORDS([thread-safe])
-AT_COMPILE_CHECK([simple/tls1.c], [], [], [], [-lpthread])
+AT_COMPILE_CHECK([simple/tls1.c], [], [], [], [-lpthread $(${PKG_CONFIG} --libs arpack)])
 AT_CLEANUP
 
 AT_SETUP([Thread-safe ARPACK:])
 AT_KEYWORDS([thread-safe ARPACK])
 AT_COMPILE_CHECK([simple/tls2.c], [simple/tls2.out], [], [internal], 
-                 [-lpthread])
+                 [-lpthread $(${PKG_CONFIG} --libs arpack)])
 AT_CLEANUP
```

**Comment by vtraag** [comment]
Thanks for thinking along with the maintance of `igraph`! Unfortunately we need to have SuiteSparse also available on Windows (compiled using MSVC) for which there are unfortunately no easy alternatives. There are some possibilities: (1) using `vcpkg` for installing SuiteSparse; (2) using pre-compiled packages from `conda`. Unfortunately, I have not managed to get either solution to work well. Feel free to experiment with both options however. If you have any solution, I would be happy to learn about it!

**Comment by vtraag** [comment]
Note by the way that we are transitioning towards changing the build system to CMake (currently under development in the `feature/cmake-build` branch).

**Comment by szhorvat** [comment]
Before switching to an external version of a library, it is important to check what (if any) changes were made to the internal version. In some cases, it is not possible to achieve full integration without modifying the library. A typical issue is the random number generator: if library should always use igraph's own generator. There may be several other things I didn't think of.

I don't know if there are any such issues with cxsparse, but we must check before moving to an external version.

A quick search reveals modifications to the CHOLMOD directory, see `CHOLMOD.diff`. These appear to be related to using the internal BLAS or an external BLAS (didn't check the details).

Another modification is to `cs/cs_randperm.c`, which is precisely to use igraph's RNG. We need to check if this function is used at all by any igraph function, or whether it can be ignored. Even if it is not currently used, a note should be made so it wouldn't be used in the future either.

**Comment by epsilon-0** [comment]
@vtraag 
Thanks a lot, I'll await the cmake release and see how it goes.

@szhorvat 
Yes, you are correct. I assumed that running the tests when building with an external cxsparse should be enough for proof of concept.
If the tests are not enough, then it is a bit beyond me to check the intricate internals and try to understand them.
I'd love to know any better methods, for me to help, that are possible for a layperson who is just managing the package without diving too much into the internals.

**Comment by szhorvat** [comment]
I should have made it clear that we *do* want to make it possible to use an external SuiteSparse. Thanks for taking the time to put this solution together!

My message above may have made the impression that I am arguing against doing it—I am not, I am merely making notes about what we need to pay attention to.

**Comment by epsilon-0** [comment]
Ah, my apologies, I misinterpreted.

Your idea sounds good, I can run tests using any patches to help if needed, but I am also busy with upgrading our whole science ecosystem at Gentoo, so my response might be a bit slow.

**Comment by szhorvat** [comment]
Noting that this is already supported on the `cmake-build` branch, so it will be available in igraph 0.9.

**Comment by ntamas** [comment]
Closing this as the request is now implemented in the `develop` branch.
